README.md
# music-stats scripts

  [![license][license-image]][license-url]
  ![code size][code-size-image]

SHOULD become an API gateway between different front-ends and various data providers, currently a set of scripts.

## Tech stack
dev deps:
[`typescript`](https://www.typescriptlang.org/docs),
[`jest`](https://jestjs.io/docs/en/expect).

deps:
[`node`](https://nodejs.org/dist/latest/docs/api) (at least v10, since experimental [`fs.promises` API](https://nodejs.org/dist/latest/docs/api/fs.html#fs_fs_promises_api) is used),
[`ramda`](http://ramdajs.com/docs),
[`axios`](https://github.com/axios/axios).

deps to consider for the server-side application:
[`koa`](http://koajs.com/#application).

## APIs, datasets
### In use
- [x] [GeoJSON Regions](https://geojson-maps.ash.ms/) (underlying source: [Natural Earth](http://naturalearthdata.com/))
- [x] [last.fm](https://www.last.fm/api/intro)
  - [x] [`library.getArtists`](https://www.last.fm/api/show/library.getArtists)
  - [x] [`user.getRecentTracks`](https://www.last.fm/api/show/user.getRecentTracks)
  - [ ] [`user.getTopArtists`](https://www.last.fm/api/show/user.getTopArtists) (pagination is fine)
  - [ ] [`user.getArtistTracks`](https://www.last.fm/api/show/user.getArtistTracks) (pagination seems to be weird, always giving `"totalPages": "0"`)
  - [ ] [`artist.getInfo`](https://www.last.fm/api/show/artist.getInfo) and [`track.getInfo`](https://www.last.fm/api/show/track.getInfo) (there are also `artist.getTags` and `track.getTags` endpoints but those simply return lists of tag names and URLs, while `.getInfo` also supplies tags plus additional data, e.g. track duration)
- [x] [MusicBrainz](https://musicbrainz.org/doc/Development/XML_Web_Service/Version_2)
- [x] [ISO 3166-1 alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) country codes as [JSON](https://gist.github.com/ssskip/5a94bfcd2835bf1dea52)

### To consider
- [ ] [Spotify](https://developer.spotify.com/documentation/web-api/reference/)
- [ ] [Songkick](https://www.songkick.com/developer/upcoming-events)
- [ ] [FMA (Free Music Archive)](https://github.com/mdeff/fma)
- [ ] [DuckDuckGo Instant Answers](https://duckduckgo.com/api)

On Spotify, only [Personalization API](https://developer.spotify.com/documentation/web-api/reference/personalization)
is available now (among other endpoints, but that's the only section about the user's listening habits),
restricting to top 50 artists/tracks. And it doesn't provide any measurable metric except "popularity"
which is some abstract (i.e. calculated) affinity level. Geo data (e.g. country) is also not there.

## Setup
### Environment variables
Create a `.env` file and fill its values according to [`.env.template`](.env.template):
* `LASTFM_API_KEY` (see last.fm [docs](https://www.last.fm/api/authentication))

### Commands
```bash
$ npm i                # install deps
$ npm run lint         # lint scripts
$ npm test             # run unit tests
$ npm run build        # compile TypeScript
$ npm run build:watch  # compile with watch
```

## Terminology
In this repository, locations are called *areas*, not *countries* (like in [music-stats/map](https://github.com/music-stats/map)).
Even though country codes are used in the resulting `merged-artists` dataset, name *area* was chosen because MusicBrainz
uses it in their schema - it can represent countries, regions or cities.

## Scripts
### Artist-area map
#### Fetch top artists for a given last.fm user
```bash
$ npm run script:artist-area-map:1-fetch-artists [50] [--no-color] [--no-cache]
#                                                 ^
#                                                 number of artists, default is set in the config
```

##### Input
Username is set in `src/config.ts`.

##### Output
Filename: `output/artist-area-map/1-lastfm-user-library.json`.

Content:
```js
[ { name: 'Dream Theater',
    playcount: 769,
    mbid: '28503ab7-8bf2-4666-a7bd-2644bfc7cb1d' }, // MusicBrainz ID
  { name: 'Queen',
    playcount: 757,
    mbid: '420ca290-76c5-41af-999e-564d7c71f1a7' },
  ...
  { name: 'Обійми Дощу',
    playcount: 222,
    mbid: 'fdafffec-3f14-442b-9700-1b52b89351ed' },
  { name: 'Lake of Tears',
    playcount: 214,
    mbid: '62cfcc64-a7d2-4ec2-ab4b-2a6b62e53940' } ]
```

#### Fetch areas for a given set of artists
```bash
$ npm run script:artist-area-map:2-fetch-artists-areas [10] [--no-color] [--no-cache]
#                                                       ^
#                                                       number of artists, default is set in the config
```

##### Input
An output of `npm run script:artist-area-map:1-fetch-artists`.

##### Output
Filename: `output/artist-area-map/2-musicbrainz-artists-areas.json`.

Content:
```js
[ { artist: 'Dream Theater', area: 'New York' }, // New York will be mapped to United States, individual cities aren't supported
  { artist: 'Queen', area: 'Japan' }, // Japan? "mbid" received from last.fm must be wrong, area will be switched to United Kingdom
  ...
  { artist: 'Обійми Дощу', area: 'Ukraine' },
  { artist: 'Lake of Tears', area: 'Sweden' } ]
```

#### Merge results of two scripts above
```bash
$ npm run script:artist-area-map:3-merge-artists [--no-color]
```

##### Input
Expects both input files (`.json`) to be located at `output/artist-area-map/`.
Blends them together, applies three stages of corrections (see `data/corrections/`),
sorts alphabetically by artist name and puts "ISO 3166-1 alpha-2" country codes as area names.

Each `[artist, playcount, countryCode]` entry is placed on a separate line to make diffs easier to digest.

##### Output
Filename: `output/artist-area-map/3-merged-artists.json`.

Content:
```js
[ [ 'Dream Theater', 769, 'US' ],
  [ 'Lake of Tears', 214, 'SE' ],
  [ 'Queen', 757, 'GB' ],
  [ 'Обійми Дощу', 222, 'UA' ],
  ... ]
```

#### Prepare country borders GeoJSON
* Filters out countries not mentioned in the merged artists dataset.
* Trims unused properties from the "GeoJSON Regions" dataset.
```bash
$ npm run script:artist-area-map:4-trim-world-map [--no-color]
```

##### Input
* An output of `npm run script:artist-area-map:3-merge-artists`.
* A GeoJSON file with country borders is located at `input/world.geo.json`.

##### Output
Filename: `output/artist-area-map/4-world.geo.json`.

### Scrobble timeline
#### Fetch all scrobbles
```bash
$ npm run script:scrobble-timeline:1-fetch-scrobbles [2019-02-25] [2019-03-10] [--no-color] [--no-cache]
#                                                     ^            ^
#                                                     |            date to (YYYY-MM-DD), defaults to today
#                                                     |
#                                                     date from (YYYY-MM-DD), defaults to yesterday
```

##### Input
None.

##### Output
Filename: `output/scrobble-timeline/1-scrobbles/2019-02-25--2019-03-10.json`.

Content:
```js
[ [ '2019-03-10 18:13',       // date
    'Handle This',            // track name
    2,                        // track playcount
    'All Killer, No Filler',  // album name
    28,                       // album playcount
    'Sum 41',                 // artist name
    33 ],                     // artist playcount
  ... ]
```

#### Merge all fetched scrobbles together
```bash
$ npm run script:scrobble-timeline:2-merge-scrobbles [--no-color]
```

##### Input
Expects input files (`.json`) to be located at `output/scrobble-timeline/1-scrobbles/`.

##### Output
Filename: `output/scrobble-timeline/2-merged-scrobbles.json`.

Content: same as from the fetching commands, but everything put into a single chronological collection with playcount values aggregated.

[license-image]: https://img.shields.io/github/license/music-stats/scripts.svg?style=flat-square
[license-url]: https://github.com/music-stats/scripts/blob/master/LICENSE
[code-size-image]: https://img.shields.io/github/languages/code-size/music-stats/scripts.svg?style=flat-square


CHANGELOG.md
# Changelog

All notable changes to this project are documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.0] - 2021-01-09
### Changed
#### Scrobble timeline
* [breaking] Dataset format update: fetched scrobbles list is now an array of arrays (`[[date, trackName, trackPlaycount, albumName, albumPlaycount, artistName, artistPlaycount], ...]`), not an array of objects.
* Fixed: timezone offsets were applied incorrectly.

## [1.3.0] - 2020-01-19
### Added
#### Scrobble timeline
* A dataset that maps artists to their main genres.
* A script that converts this dataset to a single `.json` file.

### Changed
#### Scrobble timeline
* Artist names are corrected (duplicates are merged).

## [1.2.0] - 2019-11-09
### Added
#### Artist-area map
A script that removes unused properties from world map GeoJSON and filters out countries not mentioned in the resulting playcount dataset.
```bash
$ npm run script:artist-area-map:4-trim-world-map
```

## [1.1.0] - 2019-11-03
### Added
#### Scrobble timeline
Merging script for imported scrobbles. Corrections are not applied yet.
```bash
$ npm run script:scrobble-timeline:2-merge-scrobbles
```

## [1.0.0] - 2019-10-13
### Changed
#### Artist-area map
* [breaking] Dataset format update: `[[artist, playcount, countryCode], ...]` (with "ISO 3166-1 alpha-2" country codes).

## [0.0.1] - 2019-08-23
### Added
Added the CHANGELOG.
Current list of features:

#### Artist-area map
Established, mainly deps and corrections receive updates.
```bash
$ npm run script:artist-area-map:1-fetch-artists
$ npm run script:artist-area-map:2-fetch-artists-areas
$ npm run script:artist-area-map:3-merge-artists
```

#### Scrobble timeline
In development, data format and API may change.
```bash
$ npm run script:scrobble-timeline:1-fetch-scrobbles
```


src/config.ts
import * as path from 'path';

import Config, {CorrectionDataType} from 'src/types/config';

const rootDir = path.resolve(__dirname, '../../');
const correctionsDir = path.resolve(rootDir, 'data/corrections/');
const datasetsDir = path.resolve(rootDir, 'data/datasets/');
const cacheDir = path.resolve(rootDir, 'cache/');
const inputDir = path.resolve(rootDir, 'input/');
const outputDir = path.resolve(rootDir, 'output/');
const artistAreaMapScriptOutputDir = path.resolve(outputDir, 'artist-area-map/');
const scrobbleTimelineScriptOutputDir = path.resolve(outputDir, 'scrobble-timeline/');

const config: Config = {
  userAgent: 'music-stats/2.0.0 (https://github.com/music-stats/scripts/)',

  connectors: {
    lastfm: {
      api: {
        root: 'https://ws.audioscrobbler.com/2.0',
      },
      username: 'markhovskiy',
      artists: {
        maxPageNumber: 12, // @see: https://www.last.fm/api/tos
        perPage: 200,
        countDefault: 50,
      },
      recentTracks: {
        maxPageNumber: 100,
        perPage: 200,
      },
      cache: {
        ttl: 7 * 24 * 60 * 60 * 1000, // 1 week
        dir: path.resolve(cacheDir, 'lastfm/'),
      },
    },

    musicbrainz: {
      api: {
        root: 'https://musicbrainz.org/ws/2',

        // In theory, MUST be less than 50 requests per second.
        // @see: https://wiki.musicbrainz.org/XML_Web_Service/Rate_Limiting
        requestFrequency: 600,
      },
      artists: {
        countDefault: 10,
      },
      cache: {
        ttl: 2 * 365 * 24 * 60 * 60 * 1000, // 2 years
        dir: path.resolve(cacheDir, 'musicbrainz/'),
      },
    },
  },

  scripts: {
    artistAreaMap: {
      fetchArtist: {
        outputFilePath: path.resolve(artistAreaMapScriptOutputDir, '1-lastfm-user-library.json'),
      },

      fetchArtistsAreas: {
        outputFilePath: path.resolve(artistAreaMapScriptOutputDir, '2-musicbrainz-artists-areas.json'),
      },

      mergeArtists: {
        corrections: {
          artistName: {
            dataType: CorrectionDataType.JsonFile,
            filePath: path.resolve(correctionsDir, '1-artist-name.json'),
          },
          artistArea: {
            dataType: CorrectionDataType.TxtFolder,
            filePath: path.resolve(correctionsDir, '2-artist-area/'),
          },
          area: {
            dataType: CorrectionDataType.TxtFolder,
            filePath: path.resolve(correctionsDir, '3-area/'),
          },
        },

        outputFilePath: path.resolve(artistAreaMapScriptOutputDir, '3-merged-artists.json'),
      },

      trimWorldMap: {
        inputFilePath: path.resolve(inputDir, 'world.geo.json'),
        outputFilePath: path.resolve(artistAreaMapScriptOutputDir, '4-world.geo.json'),
      },
    },

    scrobbleTimeline: {
      fetchScrobbles: {
        corrections: {
          artistName: {
            dataType: CorrectionDataType.JsonFile,
            filePath: path.resolve(correctionsDir, '1-artist-name.json'),
          },
        },

        outputFilePath: path.resolve(scrobbleTimelineScriptOutputDir, '1-scrobbles/', '<from>--<to>.json'),
      },

      mergeScrobbles: {
        outputFilePath: path.resolve(scrobbleTimelineScriptOutputDir, '2-merged-scrobbles.json'),
      },

      groupArtistsByGenres: {
        inputFilePath: path.resolve(datasetsDir, 'artist-genre/'),
        outputFilePath: path.resolve(scrobbleTimelineScriptOutputDir, '3-artists-by-genres.json'),
      },
    },
  },

  datasets: {
    countryCodes: {
      filePath: path.resolve(datasetsDir, 'country-codes-iso-3166-1-alpha-2.json'),
    }
  },
};

export default config;
